### PR TITLE
More, Better XML Macros

### PR DESCRIPTION
**Pull request by jmchilton**
There are two primary improvements here.
1. XML macros are now parameterizable. This should allow for generic macros for things like data tables. e88b73e
2. Galaxy's `job_conf.xml` now expands macros during loading. This should allow building macros that allow defining categories for destinations for instance and should result in simplifications and duplication reductions for complex job conf files. 483acd5

I think together these changes represent a more generic solution to the problem with Docker volumes being addressed with Intel-HSS/galaxy#28. These should allow building a generic destination for various Intel clusters with simple parameters for the things that change between destination (maybe just resource allocations).

---

More on paramerization of XML macros:

This comes in two modes - simple lists of required attributes and per-attribute parameters that may specify defaults.

The following toy example:

```
<tool>
    <expand macro="inputs" bar="hello" />
    <macros>
        <xml name="inputs" tokens="bar" token_quote="$$">
            <inputs type="the type is $$BAR$$" />
        </xml>
    </macros>
</tool>
```

will evaluate to `<tool><inputs type="the type is hello" /><tool>`. The XML macro definition tag can now take in the new attribute `tokens` - which will be interpreted as a list of parameter names. Each parameter can be specified when expanding the macros as simple attributes specifying the value of the parameter. The parameters will be wrapped in `token_quote` when creating actual tokens for replacement in the XML file - the default value for `token_quote` is `@`.

Parameters can also be defined individual to specify default values. So the following example:

```
<tool>
    <expand macro="inputs" foo="hello" />
    <expand macro="inputs" foo="world" />
    <expand macro="inputs" />
    <macros>
        <xml name="inputs" token_foo="the_default">
            <inputs>@FOO@</inputs>
        </xml>
    </macros>
</tool>
```

Would evaluate to `<tool><inputs>hello</inputs><inputs>world</inputs><inputs>the_default</inputs></tool>`.

### CODE REVIEW

**Comment by hexylena** [comment]
This is wonderful, :+1:/10, I am very excited for this.

Any reason for supporting both "dollar" and "at" parameters?

**Comment by martenson** [comment]
Looks great John. Please update our wiki to reflect these changes after merge. In case you resist I volunteer to update the wiki with your commit messages.

**Comment by jmchilton** [comment]
@erasche My intuition tells me that if we only picked one there would crop up a situation where that choice would be awkward - though I cannot think of many specific contexts where @ symbols are used like this. So I might be keen to drop the `dollar_` and simplify things to just `tokens` and `foo_token` - but these tokens operate different than the `<token>` tag which doesn't modify the name or append/prepend things - so I like being explicit with the names `at_tokens` and `at_token_foo`. So if we have `at` tokens it seems very easy to have other kinds as well.

One way in which I would probably use the two token types - is to stick with `@` for global tokens (which it seems like most people do) and then use something else for the contextual tokens.

Though you know what - `$`s are very commonly used in both cheetah and shell scripts - so that is a terrible choice for a second token type. I think I will redo this and drop the dollar token type and add amp and percent token types - unless people think we should just stick with at tokens for now.

**Comment by nsoranzo** [comment]
What about adding a `token_quote` attribute to the `<xml>` element, with `@` as default?

This would apply to all tokens in this element, which can be defined with a simple `tokens` attribute.

**Comment by jmchilton** [comment]
@nsoranzo - yes this! It is very obvious this is the right thing to do now that you explain it :).

**Comment by jmchilton** [comment]
Okay - requested modifications made. Pull request message and test cases updated.

Any remaining concerns?

**Comment by bgruening** [comment]
This will change a lot! Thanks John! :+1:

**Comment by martenson** [comment]
:+1:

**Comment by davebx** [comment]
:+1:

**Comment by nsoranzo** [comment]
Can you add the new `lib/galaxy/util/xml_macros.py` to `.ci/pep8_sources.txt` ?

File: test/unit/tools/test_tool_loader.py [hunk 1]
```diff
@@ -189,3 +189,41 @@ def load(self, name="tool.xml", preprocess=True):
         tag_el = xml.find("another").find("tag")
         value = tag_el.get('value')
         assert value == "The value.", value
+
+    # Test macros XML macros with $ expansions in attributes
```
Comment (nsoranzo) [inline test/unit/tools/test_tool_loader.py]: s/$/$$/ or change the code below.

**Comment by jmchilton** [comment]
@nsoranzo - done. Thanks for the catch. I suppose we should update CONTRIBUTING.md to mention that all new Python files should be added to `pep8_sources.txt`.

**Comment by dannon** [comment]
Asked on IRC as an informal query, but it never got a response.  Why move everything to macros?  There's non-macro related stuff that's getting moved to that module like load_tool, etc.

**Comment by jmchilton** [comment]
@dannon `load_tool` is renamed to `load` in the new module. It has no "tool" specific logic per-se except arguably the existence of template macros - which unlike "xml" and "token" macros have no use in job confs currently. If we opt-ed to templatize more of job_conf.xml though - we certainly could allow template macros in the future for consistency.

**Comment by nsoranzo** [comment]
Maybe add an example of macro use to `config/job_conf.xml.sample_advanced` ?

**Comment by dannon** [comment]
Right, I see that it was renamed from load_tool to load.  I don't particularly like this new organization at all since it seems like we've just shoved everything into a macros module when a lot of it seems higher level than macro interpretation, but I'm a -0, so that's fine.

**Comment by jmchilton** [comment]
@dannon - Are there specific things (line numbers, functions, etc...) you feel are not related to xml_macros in the new `xml_macros.py`? Would you like `imported_macro_paths` and `template_macro_params` back in `loader.py`?

**Comment by jmchilton** [comment]
@nsoranzo Example added - and added a test case since the job configuration unit tests are based on that file anyway.

File: test/unit/jobs/test_job_configuration.py [hunk 2]
```diff
@@ -128,6 +128,12 @@ def test_env_parsing( self ):
 
         assert env_dest.env[ 3 ][ "execute" ] == "module load javastuff/2.10"
 
+    def test_macro_expansion( self ):
+        self.__with_advanced_config()
+        for name in ["foo_small", "foo_medium", "foo_large", "foo_longrunning"]:
+            print self.job_config.destinations.keys()
```
Comment (nsoranzo) [inline test/unit/jobs/test_job_configuration.py]: Debug statement? I suppose it can be removed.

**Comment by nsoranzo** [comment]
> @nsoranzo Example added - and added a test case since the job configuration unit tests are based on that file anyway.

@jmchilton Awesome! Thanks for adding both!

File: test/unit/jobs/test_job_configuration.py (hunk 2 above)
Comment (jmchilton) [inline test/unit/jobs/test_job_configuration.py]: @nsoranzo fixed. Added that test file to pep8_sources.txt as well.

**Comment by nsoranzo** [comment]
:+1:

**Comment by dannon** [comment]
@jmchilton It's fine.  There isn't _much_ I'd pull out, but that the root 'load_tool' actually comes from the macros module seems off;  load_tool, raw_tool_xml_tree, set_children, replace, etc., should remain in the more general tool loader class IMO.  Merging anyway.

**Pull request merged**